cmd/panfs-multipart.go

```diff
@@ -73,31 +100,42 @@ func (fs *PANFSObjects) decodePartFile(name string) (partNumber int, etag string
 }
 
 // Appends parts to an appendFile sequentially.
-func (fs *PANFSObjects) backgroundAppend(ctx context.Context, bucket, object, uploadID string) {
-	fs.appendFileMapMu.Lock()
+func (fs *PANFSObjects) backgroundAppend(ctx context.Context, bucket, object, uploadID string, need_flock bool) {
 	logger.GetReqInfo(ctx).AppendTags("uploadID", uploadID)
 	bucketPath, err := fs.getBucketPanFSPath(ctx, bucket)
 	if err != nil {
 		logger.LogIf(ctx, err, logger.Application)
 		return
 	}
-
+	fs.appendFileMapMu.Lock()
 	file := fs.appendFileMap[uploadID]
 	if file == nil {
-		file = &panfsAppendFile{
-			filePath: pathJoin(bucketPath, panfsS3TmpDir, fs.nodeDataSerial, bgAppendsDirName, fmt.Sprintf("%s.%s", uploadID, mustGetUUID())),
+		file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
+		if file == nil {
+			fs.appendFileMapMu.Unlock()
+			return
 		}
 		fs.appendFileMap[uploadID] = file
 	}
 	fs.appendFileMapMu.Unlock()
 
-	file.Lock()
-	defer file.Unlock()
+	// No need to take a lock when bgAppend called from CompleteMultipartUpload which has been already taken the lock
+	if need_flock {
+		if !file.flock.TryLock() {
+			return
+		}
+		defer file.flock.Unlock()
+	}
 
-	// Since we append sequentially nextPartNumber will always be len(file.parts)+1
-	nextPartNumber := len(file.parts) + 1
-	uploadIDDir := fs.getUploadIDDir(bucketPath, bucket, object, uploadID)
+	fsMeta, err := fs.readMultipartMeta(bucketPath, object, uploadID)
+	if err != nil {
+		return
+	}
+	// Since we append sequentially nextPartNumber will always be len(fsMeta.Appended)+1
+	appendedLen := len(fsMeta.Appended)
+	nextPartNumber := appendedLen + 1
 
+	uploadIDDir := fs.getUploadIDDir(bucketPath, bucket, object, uploadID)
 	entries, err := readDir(uploadIDDir)
```

Comment:
I think, we have to wrap this readdir to some kind of loop as we added "trylock" logic for lock file above. As I got the idea, our intent is to prevent other background append goroutines (local or remote) from waiting on file lock if there is append in progress. So I see that background append can just exit if trylock fails to acquire lock. However, I think, we can have scenarios like this:

| Thread 1 | Thread 2 |
|---|---|
| Finish part uploading, rename part from tmp to multipart dir |  |
| Start background append |  |
| Take file lock |  |
| Read directory |  |
| Appending parts... | Finish part uploading, rename part from tmp to multipart dir |
|  | Start background append |
|  | Failed to take file lock, exit |
| Release file lock |  |

So it seems noone will care about part added by second thread. Of course, some next background append or "complete multipart upload" will pickup it, but it seems we can do a very small optimisation like:

```
prevDirStat = stat(uploadIDDir)

while true:
  entries = readDir(uploadIDDir)

  for entry in entries:
     ... # do append, update cursor to current entry

  if nothing was appended:
     # If we just iterated all entries and found that nothing
     # needs to be appended to target file, we can just
     # break upper-level loop.
     break

  # Check whether new part were added while we were uploading
  # existing ones.
  dirStat = stat(uploadIDDir)
  dirStat.mtime == prevDirStat.mtime:
     # multipart directory content was not changed,
     # we appended all current parts and can exit
     break

  prevStat = dirStat
  # Go to next iteration of re-reading directory entries.
  # If we found that directory content was not actually changed or
  # somehow broken, we will break the re-read loop.
```

Please note, that there is also possibility that new part can be added to multipart directory between _stat_ and _unlock_, but this window is much smaller.